### PR TITLE
[New Scheduler]Simplify TaskRequest

### PR DESCRIPTION
**Pull request by lixin-wei**
## Why are these changes needed?

As mentioned in https://github.com/ray-project/ray/pull/16155#issuecomment-854651876, we want to simplify the TaskRequest struct. This PR does the following refactoration:

before:
```cpp
struct ResourceRequest {
  FixedPoint demand;
};

struct ResourceRequestWithId : ResourceRequest {
  int64_t id;
};

class TaskRequest {
  std::vector<ResourceRequest> predefined_resources;
  std::vector<ResourceRequestWithId> custom_resources;
};
```

after:
```cpp
class TaskRequest {
  std::vector<FixedPoint> predefined_resources;
  unordered_map<int64_t, FixedPoint> custom_resources;
};
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by wuisawesome — APPROVED** [review]

**Comment by ericl** [comment]
Actually if I wanted to add scheduling locality hints (like pack vs spread), would we want to keep this type?

**Comment by rkooo567** [comment]
nice change :)

**Comment by wuisawesome** [comment]
@ericl Yes this would be the place to do it. That would be an example of why ResourceSet and TaskRequest are different, but I think this change is fine either way.

**Comment by ericl** [comment]
We're planning to add that kind of thing soon, so does it makes sense to not merge this PR then?

**Comment by wuisawesome** [comment]
I think we should merge this, but think twice before merging #16155

**Comment by rkooo567** [comment]
Hmm if we need structure for locality hints, we shouldn’t merge it. We might support this in a couple sprint ideally. I can discuss offline this to you?

**Comment by clarkzinzow** [comment]
+1 on keeping the existing structure if it will make supporting locality hints easier.

**Comment by wuisawesome** [comment]
K we're all in agreement that this PR will make future scheduling development easier. test_td3 is a known flaky test, so merging.

**Pull request merged**
